File: patrik-insane-cryptographic-authentication/elixir/patrik-insane-cryptographic-authentication.exs
Mix.install([
  {:httpoison, "~> 2.0"},
  {:jason, "~> 1.4"}
])

defmodule PatrikInsaneCryptographicAuthentication do
  @alphabet "abcdefghijklmnopqrstuvwxyz"
  @username "igor"

  def run([filename]) do
    File.stream!(filename)
    |> Stream.map(&String.trim/1)
    |> Stream.map(&String.split(&1, ":"))
    |> Stream.map(&authenticate/1)
    |> Stream.map(&greetingToken/1)
    |> Stream.map(&greeting/1)
    |> Enum.each(&IO.puts/1)
  end

  def run(_) do
    IO.puts("Usage: elixir patrik-insane-cryptographic-authentication.exs <filename>")
  end

  defp authenticate([username, password]) do
    url = "https://code-club-website.patrik-dvoracek.workers.dev/rest/puzzle/2025/september/auth"
    body = %{type: "credentials", username: username, password: password}
    json_body = Jason.encode!(body)

    headers = [
      "Content-Type": "application/json",
      "X-Cookie-User": @username
    ]

    case HTTPoison.post(url, json_body, headers) do
      {:ok, %HTTPoison.Response{status_code: _, body: response_body}} ->
        case Jason.decode(response_body) do
          {:ok, %{"token" => token, "type" => "greeting"}} ->
            token
        end
    end
  end

  defp greetingToken(token) do
    token
    |> String.graphemes()
    |> Enum.reduce(100, fn char, total ->
      lowercase_char = String.downcase(char)

      case :binary.match(@alphabet, lowercase_char) do
        :nomatch ->
          total

        {index, _} ->
          <<ascii_value::utf8>> = lowercase_char
          rem((total + index + 1) * ascii_value, 4096)
      end
    end)
  end

  defp greeting(token) do
    url = "https://code-club-website.patrik-dvoracek.workers.dev/rest/puzzle/2025/september/auth"
    body = %{type: "greeting", token: token}
    json_body = Jason.encode!(body)

    headers = [
      "Content-Type": "application/json",
      "X-Cookie-User": @username
    ]

    case HTTPoison.post(url, json_body, headers) do
      {:ok, %HTTPoison.Response{status_code: _, body: response_body}} ->
        case Jason.decode(response_body) do
          {:ok, %{"result" => result, "type" => "authenticated"}} ->
            result
        end
    end
  end
end

PatrikInsaneCryptographicAuthentication.run(System.argv())

File: patrik-insane-cryptographic-authentication/swift/patrik-insane-cryptographic-authentication.swift
import Foundation

let arguments = CommandLine.arguments

let filePath: String
switch arguments.count {
case 2:
    filePath = arguments[1]
default:
    print("Usage: swift patrik-insane-cryptographic-authentication.swift <file>")
    exit(0)
}

let url = URL(string: "https://code-club-website.patrik-dvoracek.workers.dev/rest/puzzle/2025/september/auth")!
let content = try String(contentsOfFile: filePath, encoding: .utf8).trimmingCharacters(in: .whitespacesAndNewlines)
let parts = content.split(separator: ":")
let username = parts[0]
let password = parts[1]

func createRequest<T: Encodable>(body: T) throws -> URLRequest {
    var request = URLRequest(url: url)
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue("igor", forHTTPHeaderField: "X-Cookie-User")
    request.httpMethod = "POST"
    request.httpBody = try JSONEncoder().encode(body)
    return request
}

struct AuthRequest: Encodable {
    let username: String
    let password: String
    let type = "credentials"
}

struct AuthResponse: Decodable {
    let token: String
}

let (authData, _) = try await URLSession.shared.data(for: createRequest(body: AuthRequest(username: String(username), password: String(password))))
let token = try JSONDecoder().decode(AuthResponse.self, from: authData).token

let alphabet = "abcdefghijklmnopqrstuvwxyz"
let greetingToken = token.reduce(100) { sum, char in
    let lowercased = Character(char.lowercased())
    guard let index = alphabet.firstIndex(of: lowercased) else {
        return sum
    }
    let indexValue = alphabet.distance(from: alphabet.startIndex, to: index)
    return ((sum + indexValue + 1) * Int(lowercased.asciiValue!)) % 4096
}

struct GreetingRequest: Encodable {
    let token: Int
    let type = "greeting"
}

struct GreetingResponse: Decodable {
    let result: String
}

let (greetingData, _) = try await URLSession.shared.data(for: createRequest(body: GreetingRequest(token: greetingToken)))
let result = try JSONDecoder().decode(GreetingResponse.self, from: greetingData).result
print(result)
